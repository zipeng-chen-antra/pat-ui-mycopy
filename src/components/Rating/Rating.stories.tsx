import * as React from 'react';
import Rating from './Rating';

export default {
  title: 'Rating',
  component: Rating,
};

export const DefaultRating = () => (
  <>
    <label>
      <Rating size="small" />
    </label>
    <label>
      <Rating />
    </label>
    <label>
      <Rating size="large" />
    </label>
  </>
);

export const DisabledRating = () => (
  <>
    <Rating disabled={true} defaultRating={4} />
  </>
);

export const NoRating = () => (
  <>
    <label>
      <Rating defaultRating={0} />
    </label>
    <label>
      <Rating disabled={true} defaultRating={0} />
    </label>
  </>
);
