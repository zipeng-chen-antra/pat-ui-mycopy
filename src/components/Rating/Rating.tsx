import React, {
  FC,
  useState,
} from 'react';
import { FiStar } from 'react-icons/fi';

export type RatingSize = 'small' | 'medium' | 'large';
export type RatingVisual = 'filled' | 'empty';
export interface IRatingProps {
  /** set customized style */
  className?: string;
  /** set disabledness */
  disabled?: boolean;
  /** set label */
  label?: string;
  /** set default stars highlighted */
  defaultRating?: number;
  /** set total amount of stars */
  totalStars?: number;
  /** set size of stars */
  size?: RatingSize;
  /** set star fill empty or full */
  starFill?: RatingVisual;
  // rating?: number;
}

const Rating: FC<IRatingProps> = ({
  totalStars = 5,
  disabled = false,
  size = 'medium',
  defaultRating = 2,
  label = 'Example label',
}) => {
  const [rating, setRating] = useState(defaultRating);
  const [hover, setHover] = useState(0);
  return (
    <>
      <label>{!label ? size : label}</label>
      <span className="star-container" data-testid="star-container">
        {[...Array(totalStars)].map((star, i) => {
          const inputValue = i + 1;
          return (
            <label className="star-label" key={i}>
              <input
                data-testid={`star-input-${inputValue}`}
                type="radio"
                className="star-input"
                name="rating"
                value={inputValue}
                onClick={() => !disabled && setRating(inputValue)}
              ></input>
              <FiStar
                className={`star-${size} ${disabled && `star-disabled`}`}
                color={inputValue <= (hover || rating) ? 'yellow' : 'grey'}
                size={size}
                onMouseEnter={() => !disabled && setHover(inputValue)}
                onMouseLeave={() => !disabled && setHover(0)}
              ></FiStar>
            </label>
          );
        })}
      </span>
    </>
  );
};

export default Rating;
