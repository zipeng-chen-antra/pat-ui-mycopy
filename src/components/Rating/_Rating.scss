.star-input{
    visibility:hidden;
}

.star{

    &-small{
        width: 30px;
    }
    &-medium{
        width: 50px;
    }
    &-large{
        width: 70px;
    }
    &-disabled{
        cursor: not-allowed;
        color:gray;
        opacity: .3;
    }
}


.star-container{
    display: flex;
}