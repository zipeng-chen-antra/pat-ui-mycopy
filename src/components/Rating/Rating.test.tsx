import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import Rating, { IRatingProps } from './Rating';

describe('Rating Component', () => {
  it('should render default Rating component', () => {
    const { asFragment } = render(<Rating />);
    
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Rating component with different values based on props', () => {
    const ratingProps: IRatingProps = {
      defaultRating: 1,
      size: 'large',
    };
    const component = render(<Rating {... ratingProps} />);
    const rating = screen.queryByText(
      'Normal large'
    ) as HTMLElement;
    const element = screen.getByTestId('star-container')
    expect(element).toBeInTheDocument();
    const starInput = screen.getAllByRole('radio');
    expect(+starInput[0].value).toBe(1);
  });

  it('should render Rating component disabled', () => {
    const ratingProps: IRatingProps = {
      disabled: true,
      totalStars: 5,
    };
    const  container = render(<Rating />);
    render(<Rating {...ratingProps} />);
    const rating = screen.queryByText('Disabled') as HTMLElement;
    expect(rating).toBeInTheDocument();
    const starInput = screen.getAllByRole('radio');
    fireEvent.click(starInput[3]);
    expect(+starInput[1].value).toBe(2);
  });

  it('should render the prop amount of stars', async () => {
    const ratingProps: IRatingProps = {
      totalStars: 8,
    };
    render(<Rating {...ratingProps} />);
    // const totalStars = await screen.findAllByText('Normal medium');
    const element = await screen.getByTestId('star-container');
    expect(element.querySelectorAll('label')).toHaveLength(8);
  });

});
