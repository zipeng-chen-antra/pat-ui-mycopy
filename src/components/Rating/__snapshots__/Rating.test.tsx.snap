// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating Component should render default Rating component 1`] = `
<DocumentFragment>
  <label>
    Example label
  </label>
  <span
    class="star-container"
    data-testid="star-container"
  >
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-1"
        name="rating"
        type="radio"
        value="1"
      />
      <svg
        class="star-medium false"
        color="yellow"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: yellow;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-2"
        name="rating"
        type="radio"
        value="2"
      />
      <svg
        class="star-medium false"
        color="yellow"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: yellow;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-3"
        name="rating"
        type="radio"
        value="3"
      />
      <svg
        class="star-medium false"
        color="grey"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: grey;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-4"
        name="rating"
        type="radio"
        value="4"
      />
      <svg
        class="star-medium false"
        color="grey"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: grey;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-5"
        name="rating"
        type="radio"
        value="5"
      />
      <svg
        class="star-medium false"
        color="grey"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: grey;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-6"
        name="rating"
        type="radio"
        value="6"
      />
      <svg
        class="star-medium false"
        color="grey"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: grey;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <label
      class="star-label"
    >
      <input
        class="star-input"
        data-testid="star-input-7"
        name="rating"
        type="radio"
        value="7"
      />
      <svg
        class="star-medium false"
        color="grey"
        fill="none"
        height="medium"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: grey;"
        viewBox="0 0 24 24"
        width="medium"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
        />
      </svg>
    </label>
    <button>
      Clear Rating
    </button>
  </span>
</DocumentFragment>
`;
